internal/inspect/policy.go

```diff
@@ -0,0 +1,315 @@
+// Copyright 2021-2024 Zenauth Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+package inspect
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/google/cel-go/common/ast"
+
+	policyv1 "github.com/cerbos/cerbos/api/genpb/cerbos/policy/v1"
+	responsev1 "github.com/cerbos/cerbos/api/genpb/cerbos/response/v1"
+	runtimev1 "github.com/cerbos/cerbos/api/genpb/cerbos/runtime/v1"
+	"github.com/cerbos/cerbos/internal/compile"
+	"github.com/cerbos/cerbos/internal/namer"
+	"github.com/cerbos/cerbos/internal/policy"
+)
+
+func Policies() *Policy {
+	return &Policy{
+		toResolve: make(map[string]map[string]bool),
+		imports:   make(map[string][]string),
+		results:   make(map[string]*responsev1.InspectPoliciesResponse_Result),
+	}
+}
+
+type Policy struct {
+	toResolve map[string]map[string]bool
+	imports   map[string][]string
+	results   map[string]*responsev1.InspectPoliciesResponse_Result
+}
+
+// Inspect inspects the given policy and caches the inspection related information internally.
+func (pol *Policy) Inspect(p *policyv1.Policy) error {
+	if p == nil {
+		return fmt.Errorf("policy is nil")
+	}
+
+	policyID := namer.PolicyKey(p)
+	variables := policy.ListVariables(p)
+	if _, ok := p.PolicyType.(*policyv1.Policy_ExportVariables); ok {
+		sort.Slice(variables, func(i, j int) bool {
+			return variables[i].Name < variables[j].Name
+		})
+
+		if len(variables) > 0 {
+			pol.results[policyID] = &responsev1.InspectPoliciesResponse_Result{
+				Variables: variables,
+			}
+		}
+
+		return nil
+	}
+
+	referencedVariables, err := pol.inspectConditions(p)
+	if err != nil {
+		return fmt.Errorf("failed to inspect conditions of the policy %s: %w", policyID, err)
+	}
+	pol.imports[policyID] = pol.inspectImports(p)
+
+	localVariables := make(map[string]struct{})
+	for _, variable := range variables {
+		localVariables[variable.Name] = struct{}{}
+		if _, ok := referencedVariables[variable.Name]; ok {
+			variable.Used = true
+		}
+	}
+
+	toResolve := false
+	for name := range referencedVariables {
+		if _, ok := localVariables[name]; !ok {
+			if pol.toResolve[policyID] == nil {
+				pol.toResolve[policyID] = make(map[string]bool)
+			}
+
+			toResolve = true
+			pol.toResolve[policyID][name] = false
+		}
+	}
+
+	// sort variables if there is nothing to resolve since we are not going to modify variables in the future.
+	if !toResolve {
+		sort.Slice(variables, func(i, j int) bool {
+			return variables[i].Name < variables[j].Name
+		})
+	}
+
+	a := policy.ListActions(p)
+	sort.Strings(a)
+	if len(a) > 0 || len(variables) > 0 {
+		pol.results[policyID] = &responsev1.InspectPoliciesResponse_Result{
+			Actions:   a,
+			Variables: variables,
+		}
+	}
```

Comment:
```suggestion
	// sort variables if there is nothing to resolve since we are not going to modify variables in the future.
	if len(po.toResolve[policyID]) > 0 {
		sort.Slice(variables, func(i, j int) bool {
			return variables[i].Name < variables[j].Name
		})
	}

	a := policy.ListActions(p)
	sort.Strings(a)

	pol.results[policyID] = &responsev1.InspectPoliciesResponse_Result{
		Actions:   a,
		Variables: variables,
	}
```